File: apps/desktop/src/state/runbooks.ts
import { invoke } from "@tauri-apps/api/core";

export interface RunbookSummary {
  path: string;
  name: string;
  description: string | null;
  step_count: number;
}

export type StepMode = "auto" | "manual";

export interface Step {
  title: string;
  command: string;
  language: string;
  mode: StepMode;
  /** Optional shell condition from `blaze: if=…` / `unless=…`. */
  condition: string | null;
  /** True when the directive was `unless=…` (negate the exit status). */
  negate: boolean;
}

export interface Runbook {
  name: string | null;
  description: string | null;
  steps: Step[];
}

export const listRunbooks = (): Promise<RunbookSummary[]> => invoke("runbooks_list", {});
export const loadRunbook = (path: string): Promise<Runbook> => invoke("runbooks_load", { path });
export const runbooksDir = (): Promise<string | null> => invoke("runbooks_dir");

// ---- save ----

export interface SaveStep {
  title: string;
  command: string;
  language?: string | null;
}

export interface SaveArgs {
  name: string;
  description?: string | null;
  steps: SaveStep[];
  dir?: string | null;
  overwrite?: boolean;
}

export interface SaveResult {
  path: string;
  filename: string;
}

export const saveRunbook = (args: SaveArgs): Promise<SaveResult> =>
  invoke<SaveResult>("runbooks_save", { args });

File: apps/desktop/src-tauri/src/runbooks.rs
//! Tauri command surface for runbook discovery and loading.

use std::path::PathBuf;

use blaze_runbook::{
    list_runbooks, load_runbook, save_runbook, Runbook, RunbookSummary, SaveRequest, SaveResult,
    SaveStep,
};
use serde::Deserialize;

use crate::settings;

fn resolve_dir(override_dir: Option<String>) -> Option<PathBuf> {
    if let Some(dir) = override_dir {
        if !dir.is_empty() {
            return Some(PathBuf::from(dir));
        }
    }
    let cfg = settings::load();
    if let Some(d) = cfg.runbooks.dir.filter(|s| !s.is_empty()) {
        return Some(PathBuf::from(d));
    }
    dirs::document_dir().map(|p| p.join("Blaze").join("runbooks"))
}

#[tauri::command]
pub fn runbooks_list(dir: Option<String>) -> Result<Vec<RunbookSummary>, String> {
    let Some(path) = resolve_dir(dir) else {
        return Ok(Vec::new());
    };
    list_runbooks(&path).map_err(|e| e.to_string())
}

#[tauri::command]
pub fn runbooks_load(path: String) -> Result<Runbook, String> {
    load_runbook(&PathBuf::from(path)).map_err(|e| e.to_string())
}

#[tauri::command]
pub fn runbooks_dir() -> Option<String> {
    resolve_dir(None).map(|p| p.to_string_lossy().to_string())
}

#[derive(Debug, Deserialize)]
pub struct SaveArgs {
    pub name: String,
    #[serde(default)]
    pub description: Option<String>,
    pub steps: Vec<SaveStep>,
    /// Override directory; falls back to settings/default when empty.
    #[serde(default)]
    pub dir: Option<String>,
    #[serde(default)]
    pub overwrite: bool,
}

#[tauri::command]
pub fn runbooks_save(args: SaveArgs) -> Result<SaveResult, String> {
    let dir = resolve_dir(args.dir)
        .ok_or_else(|| "no runbook directory resolvable".to_string())?
        .to_string_lossy()
        .to_string();
    save_runbook(SaveRequest {
        name: args.name,
        description: args.description,
        steps: args.steps,
        dir,
        overwrite: args.overwrite,
    })
    .map_err(|e| e.to_string())
}
